Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/filter/APMLogFilter.java

```diff
@@ -146,8 +147,9 @@ public void filter(List<LogFilterRule> logFilterRuleList) {
 
         if (logFilterRuleList.size() == 1 && logFilterRuleList.get(0).getFilterType() == FilterType.CASE_ID) {
             Set<String> selection = logFilterRuleList.get(0).getPrimaryValuesInString();
+            boolean retain = logFilterRuleList.get(0).getChoice() == Choice.RETAIN;
             filteredPTraceList = pTraceList.stream()
-                    .filter(p -> selection.contains(p.getCaseId()))
+                    .filter(p -> retain ? selection.contains(p.getCaseId()) : !selection.contains(p.getCaseId()) )
```

Comment:
A shorter way to express this logic: `p -> retain == selection.contains(p.getCaseId())`